user.go

```diff
@@ -43,8 +54,9 @@ type network struct {
 
 	history map[string]uint64
 
-	lock sync.Mutex
-	conn *upstreamConn
+	lock      sync.Mutex
+	conn      *upstreamConn
+	lastError error
```

Comment:
(If this doesn't require a lock, let's move it above `lock`, alongside `history`)